### PR TITLE
wip: feat: add issue previous/next navigation shortcuts

### LINKED ISSUE #343: Left and right navigation to quickly go through issues and commits
I'm wondering if something like this has already been added, if not myself or someone else who has the time should try and add this functionality. If you've ever used a system such as Taiga, there's nice left and right navigation buttons to quickly go through issues, especially if you have a lot of them.

Currently with Github, every time you read an issue and comment / close it, you then have to go back to the entire list of issues, remember where you left off and somehow keep track of the next issue to open.

This lack of functionality has bothered me so much that I actually sent Github staff a long email about it over a year ago. They replied but since then have not added this basic functionality, for a reference this is precisely what I'm describing, if you've ever used Taiga they already have this.

github-annoyances

As well, a similar type of functionality when you want to quickly go through commits, review changes, pull requests, etc. would also be extremely helpful.

github-annoyances2

### PR DESCRIPTION
**Pull request by iamandrewluca**
Closes #343

### CODE REVIEW

File: source/features/issues-navigation.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import features from '../libs/features';
+
+/**
+ * TODO: find a way o parse this correctly, maybe ask a GitHub developer how this is parsed?
+ * ex: `is:issue is:open sort:updated-desc label:"change request" show branch`
```
Comment (fregante) [inline source/features/issues-navigation.tsx:5]: No need to parse it, you can pass it directly to the Search API (GraphQL)

Comment (iamandrewluca) [reply source/features/issues-navigation.tsx:5]: Hmmm, didn't found a way to pass query to GraphQL API.
In v3 I see that query can be passed as is.
https://developer.github.com/v3/search/#search-issues-and-pull-requests

Comment (iamandrewluca) [reply source/features/issues-navigation.tsx:5]: Ok, I need to learn better GraphQL 🙂 a search query exists

<details>
<summary>query example</summary>

```graphql
{
  search(last: 3, type: ISSUE, query: "user:sindresorhus repo:refined-github") {
    issueCount
    pageInfo {
      hasNextPage
      endCursor
    }
    edges {
      node {
        ... on Issue {
          createdAt
          title
          url
        }
      }
    }
  }
}
```

</details>

**Comment by fregante** [comment]
Feel free to post an update when you reach a working state

**Comment by iamandrewluca** [comment]
Notes:

We can't get a cursor from an issue ID

GitHub cursor is a base64 representation of `cursor:#` where `#` is the number of order for the issue in that query.

From `document.referrer` we can take `page` query and create `before` and `after` cursor (we know a page has 25 items).

`after cursor` = `prev page` * `per page`
`before cursor` = (`current page` * `per page`) + `1` 

e.g.

`document.referrer` is `?page=3&q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc`
`page` is `3`
`after` cursor will be `btoa('cursor:50')` (2 * 25 = 50) => `Y3Vyc29yOjUw`
`before` cursor will be `btoa('cursor:76')` ((3 * 25) + 1 = 50) => `Y3Vyc29yOjc2`

Having `before` and `after` we can query the list of issue from the where current issue is coming from.

<details>
<summary>GraphQL Example</summary>

**Query**

```graphql
query SearchIssues($query: String!) {
  search(query: $query, type: ISSUE, last:25, after: "Y3Vyc29yOjUw", before: "Y3Vyc29yOjc2") {
    nodes {
      ... on Issue {
        url
      }
    }
  }
}
```

**Variables**

```json
{
  "query": "is:issue is:open sort:updated-desc user:sindresorhus repo:refined-github"
}
```

**Results**

```json
{
  "data": {
    "search": {
      "nodes": [
        {
          "url": "https://github.com/sindresorhus/caprine/issues/1244"
        },
        {
          "url": "https://github.com/sindresorhus/caprine/issues/1344"
        },
        {
          "url": "https://github.com/sindresorhus/refined-github/issues/2746"
        },
        {
          "url": "https://github.com/sindresorhus/create-dmg/issues/53"
        },
        {
          "url": "https://github.com/sindresorhus/eslint-plugin-unicorn/issues/717"
        },
        {
          "url": "https://github.com/sindresorhus/refined-github/issues/3041"
        },
        {
          "url": "https://github.com/sindresorhus/refined-github/issues/3063"
        },
        {
          "url": "https://github.com/sindresorhus/got/issues/1197"
        },
        {
          "url": "https://github.com/sindresorhus/camelcase-keys/issues/60"
        },
        {
          "url": "https://github.com/sindresorhus/got/issues/1195"
        },
        {
          "url": "https://github.com/sindresorhus/ora/issues/143"
        },
        {
          "url": "https://github.com/sindresorhus/is-port-reachable/issues/4"
        },
        {
          "url": "https://github.com/sindresorhus/gulp-filter/issues/93"
        },
        {
          "url": "https://github.com/sindresorhus/electron-context-menu/issues/103"
        },
        {
          "url": "https://github.com/sindresorhus/gulp-mocha/issues/170"
        },
        {
          "url": "https://github.com/sindresorhus/electron-better-ipc/issues/15"
        },
        {
          "url": "https://github.com/sindresorhus/caprine/issues/1213"
        },
        {
          "url": "https://github.com/sindresorhus/ky-universal/issues/8"
        },
        {
          "url": "https://github.com/sindresorhus/refined-github/issues/2934"
        },
        {
          "url": "https://github.com/sindresorhus/got/issues/1056"
        },
        {
          "url": "https://github.com/sindresorhus/got/issues/1174"
        },
        {
          "url": "https://github.com/sindresorhus/ora/issues/146"
        },
        {
          "url": "https://github.com/sindresorhus/quick-look-plugins/issues/117"
        },
        {
          "url": "https://github.com/sindresorhus/got/issues/1137"
        },
        {
          "url": "https://github.com/sindresorhus/wallpaper/issues/66"
        }
      ]
    }
  }
}
```

</details>

**Comment by fregante** [comment]
Keep in mind that GQL variables aren't currently supported by `api.ts`, so it's easier to just inline them in the string.

> GitHub cursor is a base64 representation of `cursor:#` where `#` is the number of order for the issue in that query.

That is great to know! It could make some other features feasible

**Comment by iamandrewluca** [comment]
> Keep in mind that GQL variables aren't currently supported by api.ts, so it's easier to just inline them in the string.

Yeap, I use them as inline.

Kind of found all the info to finish this feature.
Now I'm investigating how to keep some state between pages.

**Comment by iamandrewluca** [comment]
@fregante I need some suggestions.

I see that some links on GitHub don't make a full page request.
Do you now is there some special attributes to achieve this?
I inspected but could not find something useful.

The problem I have is this.
1. User goes to e.g. `https://github.com/sindresorhus/refined-github/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc`
2. User clicks and navigates to an issue
3. `document.referrer` is a issue list and Feature is activated.
4. User navigates forward to next issue using the feature
5. A full page navigation is made.
5. `document.referrer` is a issue and Feature is not activated.

I need `document.referrer` to be same when user navigates using pagination feature.

**Comment by fregante** [comment]
> I'm investigating how to keep some state between pages.

Would `history.replaceState` work? We use it in another feature. 

> I see that some links on GitHub don't make a full page request.

That’s right, some pages are loaded via ajax. A quick solution would be to add a listener to track “the previous page”, like:

```js
{
  init: () => { previousPage = currentPage; currentPage = location.href},
  repeatOnAjaxEvenOnBackButtom: true,
}
``` 

And the. Use that instead of `document.referrer` sometimes

**Comment by iamandrewluca** [comment]
I wanted to use as shortcuts `data-hotkey` attribute

https://github.com/iamandrewluca/refined-github/commit/32e5b66b196d472435fc271c7d0debf870b2ef01#diff-04c248fadc70c7eb369ec680d1f3291eR48

I'll investigate more what can I do.

**Comment by iamandrewluca** [comment]
I can keep in cache last `document.referrer` that passes `pageDetect.isDiscussionList`, and navigate using that page filters.

The problem will be when a user is landing on an issue, that does take part of that discussion list.

I'll investigate more.

**Pull request closed without merging**